p2p/transport/webrtc/stream.go

```diff
@@ -52,34 +54,42 @@ type sendState uint8
 const (
 	sendStateSending sendState = iota
 	sendStateDataSent
+	sendStateDataReceived
 	sendStateReset
 )
 
 // Package pion detached data channel into a net.Conn
 // and then a network.MuxedStream
 type stream struct {
 	mx sync.Mutex
-	// pbio.Reader is not thread safe,
-	// and while our Read is not promised to be thread safe,
-	// we ourselves internally read from multiple routines...
-	reader pbio.Reader
+
+	// readerOnce ensures that only a single goroutine reads from the reader. Read is not threadsafe
+	// But we may need to read from reader for control messages from a different goroutine.
+	readerOnce chan struct{}
```

Comment:
Nit: Don't call it once, since that's usually used for a `sync.Once`. Alternative: `readerSem`, to make clear that it's a semaphore.